README.md

[![Arduino CI](https://github.com/RobTillaart/DMM/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/DMM/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/DMM/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/DMM/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/DMM/actions/workflows/jsoncheck.yml)
[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/DMM/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/DMM.svg?maxAge=3600)](https://github.com/RobTillaart/DMM/releases)


# DMM = Digital MultiMeter.

Arduino library for a DMM (digital multimeter) class.


## Description

DMM is a class that uses the Arduino analogRead to make measurements.
It is based upon the UNO but will work on some other boards too.

The first version only works as a voltmeter, DC only.
It is meant to be extended in the future to be a complete Digital MultiMeter class.
that includes amps ohms, diode testing etc.

(this is an old 'toy' project, wrapped into a class)


#### Related

- https://github.com/RobTillaart/ACS712 (current measurement).


## Interface

```cpp
#include "DMM.h"
```

#### Constructor

- **DMM()** constructor.

#### Configuration

- **begin(uint8_t analogPin, float volts = 5.0, uint16_t maxADC = 1023)** configuration.
- **void  setMaxVoltage(float maxVoltage)**
- **float getMaxVoltage()**
- **void  setGain(float factor = 1.0)** GAIN e.g. due to voltage divider.
A 25V to 5V divider has a factor = 5.
- **float getGain()** returns the set gain.


#### ReadVolts

One must connect GND of Arduino to the GND of the device.
```
    [A0]---- V unknown
   [GND]---- GND

or with a divider => setGain(x)

    [A0]----[divider]---- V unknown
   [GND]----[divider]---- GND

```

- **float readVolts(uint8_t times = 1)** read voltage.
Times can be set to average multiple measurements.
- **float readMilliVolts(uint8_t times = 1)** convenience wrapper.
- **float readNoise(uint8_t times = 1)** measure the noise on the wire.
Assumes connected to constant voltage e.g. GND.


#### ReadOhm

The schema for measuring resistors is based on a simple voltage divider.
See below. This schema can be extended with a selector switch that can
switch between different reference resistors.

```
   [GND] ----[ R reference ]----[A0]----[ R unknown ]----[+5V]
```

- **void setReferenceR(uint32_t ohm = 1000)** set the reference resistor.
units are Ohm, default = 1K.
- **uint32_t readOhm(uint8_t times = 1)** read ohm.
Times can be set to average multiple measurements.
- **uint32_t readKiloOhm(uint8_t times = 1)** read ohm.
Times can be set to average multiple measurements.



#### Calibration

- **float readNoise(uint8_t times = 1)** get noise level for accuracy.


## Operation

Basic operation is to connect the Arduino GND to the projects GND (or -)
and use the configured analog pin to measure (positive) voltage.

By adding a voltage divider one can measure larger voltages.
The divider factor can be set with **setGain(factor)**.

To elaborate.
- GND, VCC


## Future

#### Functional

- AMPS  (how -> measure voltage over a known resistor)
- CAPS  (how -> RC timing)
  - RC (load 5 seconds)
  - disconnect
  - measure drop rate / time in micros.
- DIODE (how -> 5V or 0V)
  - DigitalRead(INPUT PULLUP) => HIGH or LOW
  - or 0.7 volt drop from 5V


#### Must

- update documentation
- update examples
- add use of AREF
  - external or internal 1.1 Volt

#### Should

- investigate noise
- analogReference(EXTERNAL) for external voltage for build in ADC
- support external ADC e.g. ADS1115 for 16 bit resolution.



#### Could

- multichannel compare
- schema for
- scope functionality?


#### Won't




CHANGELOG.md
# Change Log DMM

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.1.2] - 2023-02-03
- add OHM measurements
- update GitHub actions
- update license 2023
- refactor
- update readme.md


## [0.1.1] - 2022-12-19
- initial release...
- renamed initial class to DMM
- add setGain() and getGain()
- add getNoise()
- add examples

## [0.1.0] - eons ago
- initial version
- started as a VoltMeter only class





test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2022-12-19
// PURPOSE: unit tests for DMM class
//          https://github.com/RobTillaart/DMM
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// ----------------------------
// assertEqual(expected, actual)
// assertNotEqual(expected, actual)
// assertLess(expected, actual)
// assertMore(expected, actual)
// assertLessOrEqual(expected, actual)
// assertMoreOrEqual(expected, actual)
// assertTrue(actual)
// assertFalse(actual)
// assertNull(actual)


#include <ArduinoUnitTests.h>

#include "Arduino.h"

#ifndef A0
#define A0  14
#endif


#include "DMM.h"


unittest_setup()
{
    fprintf(stderr, "DMM_LIB_VERSION: %s\n", (char *) DMM_LIB_VERSION);
}

unittest_teardown()
{
}


//  placeholder for now

unittest(test_first)
{
  DMM dmm;
  assertEqual(1,1);
}


unittest_main()


// -- END OF FILE --


DMM.cpp
//
//    FILE: DMM.cpp
//  AUTHOR: Rob Tillaart
// VERSION: 0.1.2
//    DATE: 2022-12-19
// PURPOSE: Library for a DMM class for Arduino UNO.
//     URL: https://github.com/RobTillaart/DMM


#include "DMM.h"


DMM::DMM()
{
}


void DMM::begin(uint8_t pin, float maxVoltage, uint16_t maxSteps)
{
  _pin         = pin;
  _maxVoltage  = maxVoltage;
  _maxSteps    = maxSteps;
  _voltageStep = _maxVoltage/_maxSteps;
}


void DMM::setMaxVoltage(float maxVoltage)
{
  _maxVoltage  = maxVoltage;
  _voltageStep = _maxVoltage/_maxSteps;
}


float DMM::getMaxVoltage()
{
  return _maxVoltage;
}


void  DMM::setGain(float factor)
{
  _gain = factor;
}


float DMM::getGain()
{
  return _gain;
}


/////////////////////////////////////////////////
//
//  READ VOLTAGE
//
float DMM::readVolts(uint8_t times)
{
  uint32_t total = 0;
  if (times == 0) 
  {
    times = 1;
  }
  for (int i = 0; i < times; i++)
  {
    total += analogRead(_pin);
  }
  float volts = (total * _voltageStep) / times;

  if (_gain != 1.0) 
  {
    volts *= _gain;
  }
  return volts;
}


float DMM::readMilliVolts(uint8_t times)
{
  return readVolts(times) * 1000.0;
}


float DMM::readNoise(uint8_t times)
{
  uint16_t  maxV = 0;
  uint16_t  minV = _maxSteps;

  if (times == 0) times = 1;
  for (int i = 0; i < times; i++)
  {
    uint16_t x = analogRead(_pin);
    if (x > maxV) maxV = x;
    if (x < minV) minV = x;
  }
  float v = ((maxV - minV) * _voltageStep) / times;

  if (_gain != 1.0) v *= _gain;
  return v;
}


/////////////////////////////////////////////////
//
//  READ OHM
//
void DMM::setReferenceR(uint32_t ohm)
{
  _ReferenceR = ohm;
}


uint32_t DMM::readOhm(uint8_t times)
{
  float voltage = readVolts(times);
  
  uint32_t R = round((_maxVoltage - voltage) * _ReferenceR / voltage);
  return R;
}


uint32_t DMM::readKiloOhm(uint8_t times)
{
  return readOhm(times) / 1000;
}


//  -- END OF FILE --




DMM.h
#pragma once
//
//    FILE: DMM.h
//  AUTHOR: Rob Tillaart
// VERSION: 0.1.2
//    DATE: 2022-12-19
// PURPOSE: Library for a DMM class for Arduino.
//     URL: https://github.com/RobTillaart/DMM


#include "Arduino.h"

#define DMM_LIB_VERSION        (F("0.1.2"))


class DMM
{
public:
  DMM();

  void begin(uint8_t pin, float maxVoltage, uint16_t maxSteps);


  //  CALIBRATION
  void  setMaxVoltage(float maxVoltage);
  float getMaxVoltage();


  //  GAIN e.g. due to voltage divider
  //  a 25V to 5V divider has a factor = 5.
  void  setGain(float factor = 1.0);
  float getGain();


  //  READ VOLTAGE
  float readVolts(uint8_t times = 1);
  float readMilliVolts(uint8_t times = 1);
  float readNoise(uint8_t times = 1);


  //  READ OHM
  void     setReferenceR(uint32_t ohm);
  uint32_t readOhm(uint8_t times = 1);
  uint32_t readKiloOhm(uint8_t times = 1); 


private:
  uint8_t  _pin         = 14;  //  A0;
  float    _maxVoltage  = 5.0;
  uint16_t _maxSteps    = 1023;
  float    _voltageStep = _maxVoltage/_maxSteps;

  float    _gain        = 1.0;
  
  uint32_t _ReferenceR  = 0;
};


//  -- END OF FILE --

